### PR TITLE
fix(core): traverse associations and apply aliases on order clauses

### LINKED ISSUE #18039: [Bug]: Incorrect SQL generation for order clause when including BelongsToMany and using minifications
## Issue Creation Checklist

- [x] I understand that my issue will be automatically closed if I don't fill in the requested information
- [x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Bug Description

When querying, the order clause gets incorrectly generated when the following conditions are met:

- minifyAliases: true
- subQuery: true
- query includes a BelongsToMany
- ordering is by a deeply nested include statement (at least 2 levels)
- the attribute you are ordering by is not the primary key

### Reproducible Example

```typescript

  it('orders by an alias when ordering by an attribute in an included subquery, filtering by a belongsToMany, and minifyAliases are enabled', async () => {
      const sequelizeMinifyAliases = Support.createSingleTestSequelizeInstance({
        logQueryParameters: true,
        benchmark: true,
        minifyAliases: true,
        define: {
          timestamps: false,
        },
      });

      const Customer = sequelizeMinifyAliases.define(
        'customer',
        {
          id: {
            type: DataTypes.INTEGER,
            primaryKey: true,
            autoIncrement: true,
          },
          name: {
            type: DataTypes.TEXT,
          },
        }, { timestamps: false },
      );

      const Route = sequelizeMinifyAliases.define(
        'route',
        {
          id: {
            type: DataTypes.INTEGER,
            primaryKey: true,
            autoIncrement: true,
          },
          date: {
            type: DataTypes.DATEONLY,
          },
        },
        { timestamps: false },
      );
      const Load = sequelizeMinifyAliases.define(
        'load', {
        id: {
          type: DataTypes.INTEGER,
          primaryKey: true,
          autoIncrement: true,
        },
        routeId: {
          type: DataTypes.INTEGER,
          references: {
            model: Route,
            key: 'id',
          },

        }
      }, { timestamps: false });

      const Delivery = sequelizeMinifyAliases.define(
        'delivery',
        {
          id: {
            type: DataTypes.INTEGER,
            primaryKey: true,
            autoIncrement: true,
          },
          loadId: {
            type: DataTypes.INTEGER,
            references: {
              model: Load,
              key: 'id',
            },
          },
          customerId: {
            type: DataTypes.INTEGER,
            references: {
              model: Customer,
              key: 'id',
            },
          },

          quantity: {
            type: DataTypes.INTEGER,
          }
        },
        { timestamps: false },
      );

      const CustomerDeliveries = sequelizeMinifyAliases.define(
        'customer_deliveries',
        {
          customerId: {
            type: DataTypes.INTEGER,
          },
          deliveryId: {
            type: DataTypes.INTEGER,
          },
        },
        { timestamps: false },
      );

      Route.hasMany(Load, { foreignKey: 'routeId' });
      Load.hasMany(Delivery, { foreignKey: 'loadId' });
      Customer.belongsToMany(Delivery, {
        through: 'customer_deliveries',
        foreignKey: 'customerId',
        otherKey: 'deliveryId',
      });

      await sequelizeMinifyAliases.sync({ force: true });

      await Route.create({ id: 1, date: '2025-12-12' });
      await Load.create({ id: 2, routeId: 1 });
      await Customer.create({ id: 4, name: 'Zyzz Aziz' });
      await Delivery.create({ id: 3, quantity: 10, loadId: 2 });
      await CustomerDeliveries.create({ customerId: 4, deliveryId: 3 });

      const delivery = (
        await Delivery.findOne({
          subQuery: false,
          include: [
            {
              model: Load,
              required: true,
              attributes: ['id'],
              include: [
                {
                  model: Route,
                  required: true,
                  attributes: ['id', 'date'],
                  where: { date: { [Op.gt]: '2020-01-01' } },
                }],
            },
            {
              model: Customer,
              required: true,
              where: { name: 'Zyzz Aziz' }
            }
          ],
          where: { quantity: { [Op.gt]: 0 } },
          order: [['load', 'route', 'date', 'ASC']],
        })
      )
      expect(delivery.load.route.date).to.equal('2025-12-12');

      const deliveryWithSubquery = (
        await Delivery.findOne({
          include: [
            {
              model: Load,
              required: true,
              attributes: ['id'],
              include: [
                {
                  model: Route,
                  required: true,
                  attributes: ['id', 'date'],
                  where: { date: { [Op.gt]: '2020-01-01' } },
                }],
            },
            {
              model: Customer,
              required: true,
              where: { id: 4 }
            }
          ],
          where: { quantity: { [Op.gt]: 0 } },
          order: [['load', 'route', 'date', 'ASC']],
          logging: console.log,
        })
      );
      expect(deliveryWithSubquery.load.route.date).to.equal('2025-12-12');

    });
```

### What do you expect to happen?

return a result in the order requested.

### What is actually happening?

missing FROM-clause entry for table "load->route"

### Environment

- Sequelize version: v7.0.0-alpha47
- Node.js version: 22/LTS
- Database & Version: Postgresql v17
- Connector library & Version: pg

## Would you be willing to resolve this issue by submitting a Pull Request?

- [x] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I will need guidance.
- [ ] No, I don't have the time, but my company or I are [supporting Sequelize through donations on OpenCollective](https://opencollective.com/sequelize).
- [ ] No, I don't have the time, and I understand that I will need to wait until someone from the community or maintainers is interested in resolving my issue.

---

_Indicate your interest in the resolution of this issue by adding the 👍 reaction. Comments such as "+1" will be removed._

### PR DESCRIPTION
**Pull request by SippieCup**
This resolves the missing FROM-clause error triggered by the new Postgres integration test while keeping existing alias behaviors intact.

## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description of Changes

`getQueryOrders` now detects when an order attributes are pulled from an included subquery. Mirroring that order into the subquery so limit/offset remain stable.

We then rewrite the main order clause to use the projected column alias (respecting `minification: true` and `as` aliases) instead of assuming non-existent table aliases like `foo->bar`. 

Added shared helpers for validating order directions and for walking association paths (with tests relying on the new _getAssociationOrderInfo). 

## List of Breaking Changes

No known breaking changes:

Closes #18039

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved ORDER BY handling for association-based ordering, including proper alias resolution for nested includes and minified aliases.

* **Tests**
  * Added integration tests covering ordering by aliased attributes within nested includes and through-table associations, with and without subqueries.

* **Documentation**
  * Removed multiple CHANGELOG files and historical release entries across the repository.

### CODE REVIEW

File: packages/core/src/abstract-dialect/query-generator.js [hunk 1]
```diff
@@ -1797,6 +1807,198 @@ export class AbstractQueryGenerator extends AbstractQueryGeneratorTypeScript {
...
+    if (direction !== undefined) {
+      if (results.length === 0) {
+        results.push([direction]);
+      } else {
+        results.at(-1).push(direction);
+      }
+    }
+
+    return results.length > 0 ? results : [[]];
+  }
```
[the remark this answers is not in the document]
Comment (SippieCup) [reply packages/core/src/abstract-dialect/query-generator.js:2000]: This requires breaking up the codebase into multiple files or rewriting the entire generator into typescript. I don't think removing this logic outside of the file is a good idea, and im not rewriting the whole thing right now.

**Comment by WikiRik** [comment]
I'm playing around with AI and while I still need to practice a lot more to get the results I'm hoping for I did want to share the review it came up with (with some pointers I gave it). https://gist.github.com/WikiRik/1c2f5eaafd43ff5dd24bdf034fb57c73

As mentioned on Slack. I am using the newest commit

**Comment by SippieCup** [comment]
Critical Issues:

1. Already Resolved, no use of `has` or using a set.
2. Already resolved
3. Array Mutaton - This is intended as the order needs to be mutated directly for further changes downstream. We should not keep it atomic in nature as it does need those modified arrays.


Q: BelongsToMany through model handling

When including a belongs to many with specific where clause requirements, you need to navigate through (potentially) hidden through models within the order filter in order to generate the correct selector, regardless of minification.

There is already a test case to exercising this path.

it does indeed handle ordering by through tables.


Q: Error handling approach (Line 1883)

Question: What happens if someone has a valid but complex order clause that doesn't match the expected patterns?

Then it is not a valid order clause. this should handle all patterns regardless of complexity with the exception of string order clauses which require manual validation as it cannot be supported through this.


Question: Should we return null instead and let the existing code handle it?

No because this has determined that an incorrect SQL query will be generated. THe issue is that it may still generate _valid_ sql queries, but not the specific one intended by the author. Failing early is the best approach here since we are protecting the user from that experience.

2. Order Entry Splitting

Q: Purpose of _splitOrderEntry

We need to split up these orders in order to correctly path them through the aliases later on. this ensures all coersion to the true column names and aliases table names are used.

What problem does breaking these apart solve? 

this solves the issue of minification as it makes it easy to parse the tables and attributes down to a minified alias in a consistent manner.

3. Subquery Order Duplication


Question: Why do we need the order in both places?

one if for minification / alias handling, and the splicing out of the literals.

Question: Does this ensure LIMIT/OFFSET work correctly with the subquery?

Yes.

### Edge Cases

1. Circular association references

    What happens if there's a circular reference in associations?
    Should we add a depth limit or cycle detection?

not an issue as it only has a depth of n-1 order clauses.



2. Polymorphic associations

    How does this handle polymorphic associations?
    Test case needed?

It should be able to handle polymorphic associations, as it is going back to aliases by the end. No additional test cases are needed imo.

     

3. Custom association options

    Associations with custom foreignKey or as options
    Multiple associations to the same model with different aliases


Both handled gracefully and correctly. thus why there is a second path within subquery to handle this particular usecase.



4. Deep nesting

    - Ordering by attributes from multiple nested levels
    - Example: ['user', 'profile', 'address', 'city', 'name', 'DESC']
    - Is there a practical limit?

Not unless there is one on SQL servers as well.

5. Paranoid models

No affect.

6. Scoped associations

Also not relevant to this discussion.


----


1. What was the exact error?
   SequelizeDatabaseError: missing FROM-clause entry for table "*ASSOICATIONNAME*->*ASSOICATIONNAME"


Before / After queries can be shared in a confidential setting.





3. Why Dual Handling in getQueryOrders?


    Lines 2554-2557: Why add association order to subQuery?
    Lines 2599-2607: Why then replace it in the main order array?
    How do these two pieces work together?



Honestly. this one is just ai slop. 

You cant access subquery tables from the main query. thus why it needs to be processed down for when a subquery is used, but the order is within the main clause. It should be broken down and correctly passed into the subquery portion of the query generator..



The rest is fairly sloppy too. the critical bug doesn't exist so its grabbing that from context, likely because it did exist on an earlier git commit. (tests would fail if it was still present).

I agree with JS docs, and can add that in soon. Unit tests are mostly there. At least within this particular scope. THe complaints it has for tests are valid, but are more around the project as a whole. 

Because of how our tests are run, the additional helper methods are essentially tested throughout every other test. the suggested tests are all redundant.  


    Single association orders with and without minifyAliases <-- literally everywhere in our tests
    Nested association orders (2-3 levels deep) <-- We have some that are 14 levels deep now.
    BelongsToMany associations with subQuery <-- added by this PR.
    Mixed order clauses (associations + direct columns) <-- A unit test that does this is what found the second bug I fixed.
    Edge cases (invalid associations, missing includes, etc.) <-- this one is valid, but would error out regardless in other locations. We really don't need to test for invalid queries on the order processing PR.


Adding these additional tests as safegaurds or finding breaking tests would be a good task for someone who wants to contribute. But a lot of it is redundant.

**Comment by WikiRik** [comment]
@SippieCup thanks for the thorough answer! I'm even more sure now that I'll need to practice a lot more on this before I use this in actual reviews

File: packages/core/test/integration/dialects/postgres/query.test.js [hunk 2]
```diff
@@ -8,6 +8,7 @@ const sinon = require('sinon');
 
 const dialect = Support.getTestDialect();
 const { DatabaseError, DataTypes } = require('@sequelize/core');
+const { Op } = require('@sequelize/core');
```
Comment (WikiRik) [inline packages/core/test/integration/dialects/postgres/query.test.js]:
```suggestion
const { DatabaseError, DataTypes, Op } = require('@sequelize/core');
```

As per coderabbit's valid suggestion

**Review by WikiRik — COMMENTED** [review]
I'm a bit on the fence on `_getAssociationOrderInfo` about finding the associations. Do we not have a helper function for that yet? If so; it might be good to make one instead of defining it all in the `_getAssociationOrderInfo` function

**Comment by WikiRik** [comment]
> I'm a bit on the fence on `_getAssociationOrderInfo` about finding the associations. Do we not have a helper function for that yet? If so; it might be good to make one instead of defining it all in the `_getAssociationOrderInfo` function

We've discussed this on Slack and we're not working on that now. We might want this in the future, but not a high priority
